### ISSUE TITLE
Replace Docker build with buildx for multi environment support?

### ISSUE THREAD

**Issue by and712** [farmOS/farmOS#339]
as stated by @mstenta in "Run on Rasperry Pi via Docker" https://github.com/farmOS/farmOS/issues/58 it is not possible to use the docker hub image on Raspberry Pis. 
Also described in the ticket is the still functional workaround (only the folders changed). 

In https://www.drupal.org/project/farm/issues/3162767 and https://www.drupal.org/project/farm/issues/3162686 @mstenta also mentions the changes for the build process for farmOS 2.0
It seems like the change to buildx pipelines with Github is easy, because there is a Action in Github: https://github.com/marketplace/actions/docker-buildx 

I never tried to build a docker image or used Github Actions bevor and got a semi working PoC in ~1h. 
Whats missing / not working: 

- publishing on docker hub (didnt try till now)
- run 'docker-entrypoint.sh' (Didnt really get what this script is doing)

Now my questions are: Is it even useful to update the build process for 1.0 or is it better to wait for 2.0 
(Probably also depends on when 2.0 is released ;) ) and is it even a good idea to use buildx? 

````
name: Test 7.x-1.x Docker and Drush Make builds
on:
  push:
    branches:
      - '7.x-1.x'

jobs:
  buildx:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout
        uses: actions/checkout@v2
      - name: Set up Docker Buildx
        id: buildx
        uses: crazy-max/ghaction-docker-buildx@v3
        with:
          buildx-version: latest
          qemu-version: latest
      - name: Available platforms
        run: echo ${{ steps.buildx.outputs.platforms }}
      - name: Set FARMOS_BRANCH environment variable
        run: echo ::set-env name=FARMOS_BRANCH::${GITHUB_REF:11}
      - name: Run Buildx
        run: |
          docker buildx build \
            --platform linux/386,linux/amd64,linux/arm/v6,linux/arm/v7,linux/arm64,linux/ppc64le,linux/s390x \
            --output "type=image,push=false" \
            --build-arg FARMOS_REPO=https://github.com/${GITHUB_REPOSITORY} \
            --build-arg FARMOS_BRANCH=${FARMOS_BRANCH} \
            --file ./docker/dev/Dockerfile ./docker/dev
````

**Comment by mstenta** [comment]
@and712 I hadn't heard of `docker buildx` until now - can you help me understand what problem(s) it would solve?

FWIW the 2.x Build Process is largely complete, and I don't anticipate changing it drastically moving forward. But, if `buildx` solves a particular problem (with Raspberry Pi builds you mentioned?) then perhaps it's still worth thinking about where it might fit.

**Comment by and712** [comment]
@mstenta 
i was a bit fast here.. 
The image builds on the raspberry pi with that Dockerfile: https://github.com/farmOS/farmOS/blob/7.x-1.x/docker/Dockerfile, but not with the new dev one (https://github.com/farmOS/farmOS/blob/7.x-1.x/docker/dev/Dockerfile), because the dev one is using the farmos image as inherit. 

Also with the old one it builds and creates a drupal site, but without the farmos distro? (i didnt had to much time looking into it). I just set up a linux vm and will investigate further (because docker on windows is.. lets say not good.. :/ ). 

Buildx could provide an image for all different platforms. I think the current way is to do this with manifest files, but buildx seems to be a bit easier: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/ 

i think providing a farmOS would be really nice and add to the security of the instances (easy updates). 

And also i can be added later ;)

**Comment by mstenta** [comment]
All very interesting - thanks for sharing @and712 .

> The image builds on the raspberry pi with that Dockerfile: https://github.com/farmOS/farmOS/blob/7.x-1.x/docker/Dockerfile, but not with the new dev one (https://github.com/farmOS/farmOS/blob/7.x-1.x/docker/dev/Dockerfile), because the dev one is using the farmos image as inherit.

Assuming you are building the images directly on the Pi, an easy workaround to this is to just build both, and edit the `FROM` line in the `dev/Dockerfile` to reference the first image you build. Or... build the first image using the name `farmos/farmos:7.x-1.x`, so that it overrides the Docker Hub tag name:

`sudo docker build -t farmos/farmos:7.x-1.x docker`

Then you don't need to edit `dev/Dockerfile`, because it will find the image tagged as `farmos/farmos:7.x-1.x` locally and won't need to download it.

> Also with the old one it builds and creates a drupal site, but without the farmos distro? (i didnt had to much time looking into it). I just set up a linux vm and will investigate further (because docker on windows is.. lets say not good.. :/ ).

Hmm I'm not sure what you mean by this. The Docker images do include the full farmOS codebase (Drupal core + farmOS distro). Unless something is happening differently on ARM builds? I'd be curious what you find when you look closer.

> Buildx could provide an image for all different platforms. I think the current way is to do this with manifest files, but buildx seems to be a bit easier: https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

Buildx does seem interesting! I'm a bit dissuaded by the "experimental" warnings I see everywhere. I wonder what the roadmap is for a stable/supported version.  I would rather wait for that before we depend upon it for critical build infrastructure.  But I fully support experimenting with this if you want to. Perhaps it would make sense to start a separate Git repository specifically for that. We could link to it from https://farmOS.org/development/projects in the meantime, and then consider merging it in the future if it seems like it will be useful for the community. What do you think?

**Comment by buckaroogeek** [comment]
For what it is worth, I am experimenting with multi-arch builds (amd64 and arm64 only at this time) using a tool called buildah.  Very early stages still but seems to work on my fedora workstation. A few very brief notes at https://hub.docker.com/r/buckaroogeek/doh-client and in the ./notes-buildah subdirectory of the supporting github project at https://hub.docker.com/r/buckaroogeek/doh-client. The ./notes-buildah/multi shell script does the work. Please note this is a naive approach and needs a lot of optimization but still illustrates the major principles.

**Comment by and712** [comment]
@mstenta I think you misunderstood me a bit. 

i tried the following thinks to build the image on the Raspberry Pi: 

## 1. use the dockerfile in docker/ 


```
git clone https://github.com/farmOS/farmOS.git
cd farmOS/docker/
sudo docker build -t farmos .
sudo docker run -p 80:80 farmos
```



### Conculusion 
Build process works but not the newest version of farmOS will be build? 
iam not sure. i see all the modules, but i cant access them. 
![buildfarm-main](https://user-images.githubusercontent.com/30082425/89910150-2ce41300-dbf0-11ea-801c-1bbc5b059829.PNG)


## 2. use the dockerfile in docker/dev 

```
git clone https://github.com/farmOS/farmOS.git
cd farmOS/docker/dev/
sudo docker build -t farmosdev .
```

### Conculusion 

`standard_init_linux.go:211: exec user process caused "exec format error"`

building is not possible because of the base image `farmos/farmos:7.x-1.x`? 


### 3. didnt have time for another try. Probably i will try use the drupal Dockerfile

**Comment by and712** [comment]
@buckaroogeek thank you for that information! Sounds even better :)

**Comment by mstenta** [comment]
@and712 can you try this?

```
git clone https://github.com/farmOS/farmOS.git
sudo docker build -t farmos/farmos:7.x-1.x farmOS/docker
sudo docker build -t farmos/farmos:dev farmOS/docker/dev
sudo docker run --name farmos -p 80:80 farmos/farmos:dev
```

What that will do is, first build the `farmos/farmos:7.x-1.x` image locally, then it will build `farmos/farmos:dev` based on that locally built image.

That *should* work to build the `dev` image for ARM.

As for the other issue - you're right it seems to be installing the "Standard" Drupal installation profile instead of the "farmOS" one. I would be curious, after you have it running with the commands above, what do you see when you run:

`sudo docker exec -it farmos ls -al /var/www/html/profiles/farm`

That should print out the contents of the `farm` installation profile. If that's empty, then something's wrong...

**Comment by raleonardo** [comment]
I bumped into a similar issue running the farmos image on my rockpi4
I think its because the container is build for amd64 when raspberrypi runs a arm64 kernel, as well as the rockpi
```
raleonardo@rockpi-4b:~/farmOS$ docker-compose up
Starting farmos_www_1 ... done
Attaching to farmos_www_1
www_1  | standard_init_linux.go:219: exec user process caused: exec format error
farmos_www_1 exited with code 1
```

```
raleonardo@rockpi-4b:~$ docker inspect farmos/farmos:dev |grep Architecture
        "Architecture": "amd64",
```

I'm running armbian:
```
raleonardo@rockpi-4b:~$ uname -rm
5.10.12-rockchip64 aarch64
```

**Comment by mstenta** [comment]
Hi @raleonardo - did you try building the image locally as described in my previous comment?

Try running this before you run `docker-compose up`:

```
git clone https://github.com/farmOS/farmOS.git
sudo docker build -t farmos/farmos:7.x-1.x farmOS/docker
sudo docker build -t farmos/farmos:dev farmOS/docker/dev
```

That will essentially build the images locally, but give them the same names as the Docker Hub images, so when your `docker-compose.yml` creates the containers it will find the image that it is looking for locally, built for ARM.

(This is just the short-term method for using farmOS on Pi right now. Ultimately it would be nice to build ARM variants and host them on Docker Hub as well!)

**Comment by raleonardo** [comment]
Hi @mstenta

Now I did..
I think i have to build it some other way.. prob with buildx.
 
```
Sending build context to Docker daemon   5.12kB
Step 1/15 : FROM farmos/farmos:7.x-1.x
 ---> 1397a8b028e6
Step 2/15 : ARG FARMOS_REPO=https://git.drupal.org/project/farm.git
 ---> Using cache
 ---> 88fbb6a630b3
Step 3/15 : ARG FARMOS_BRANCH=7.x-1.x
 ---> Using cache
 ---> 2a2f59ff6331
Step 4/15 : ARG WWW_DATA_ID=1000
 ---> Using cache
 ---> 8f5c14ab94a1
Step 5/15 : RUN usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in 3e67b6a68cd1
standard_init_linux.go:219: exec user process caused: exec format error
The command '/bin/sh -c usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data' returned a non-zero code: 1
```

Didn't research this much yet, but I guess it's due to the arch of the layers upsteam

**Comment by raleonardo** [comment]
btw, i just pulled drupal:7 (which is the FROM in your dockerfile) and ran:
`raleonardo@rockpi-4b:~$ docker pull drupal:7
7: Pulling from library/drupal
83c5cfdaa538: Pull complete 
7445693bd43e: Pull complete 
8eb0363d084b: Pull complete 
b87c24959505: Pull complete 
4d1c5d697df1: Pull complete 
361e2a373dc1: Pull complete 
df6059823413: Pull complete 
282e62e4b135: Pull complete 
4fbec33b64b9: Pull complete 
c7643dc96f99: Pull complete 
113254d1e77a: Pull complete 
8c6b49bef0eb: Pull complete 
6572473fc343: Pull complete 
84935d3a5826: Pull complete 
2073ddc34f28: Pull complete 
0c1b1d6c526e: Pull complete 
Digest: sha256:b9ff35ed45cf186556aeb695605ad2d385250574ff7455581e7e815d93a17c66
Status: Downloaded newer image for drupal:7
docker.io/library/drupal:7
raleonardo@rockpi-4b:~$ 
raleonardo@rockpi-4b:~$ docker inspect drupal:7 | grep Arch
        "Architecture": "arm64",`

**Comment by raleonardo** [comment]
So I tried building on the rockpi from the normal Dockerfile. It built (looong) and ran OK.

Not so much with the dev one: it fails to build, as the build its running on top of the layers of farmos/farmos:7.x-1.x, which were build on your pipeline for amd64. 

```
raleonardo@rockpi-4b:~/farmOS$ docker build  farmOS/docker/dev/ -t farmos/farmos:dev
Sending build context to Docker daemon   5.12kB
Step 1/15 : FROM farmos/farmos:7.x-1.x
7.x-1.x: Pulling from farmos/farmos
a076a628af6f: Pull complete 
02bab8795938: Pull complete 
657d9d2c68b9: Pull complete 
f47b5ee58e91: Pull complete 
2b62153f094c: Pull complete 
60b09083723b: Pull complete 
1701d4d0a478: Pull complete 
b058a575d643: Pull complete 
1ad503736966: Pull complete 
ae67689a4962: Pull complete 
730b1f7e463f: Pull complete 
678348961241: Pull complete 
e14e469c4fd4: Pull complete 
31a8c0b2fd1a: Pull complete 
b3c053f1c848: Pull complete 
042a6a27bc3a: Pull complete 
27ab162cfc14: Pull complete 
15131806a1cc: Pull complete 
7e6e5372eeda: Pull complete 
94a2814e476a: Pull complete 
2a6bfde8fd93: Pull complete 
f194a8a08958: Pull complete 
857c38d1308c: Pull complete 
a7f2f1463026: Pull complete 
d4887b294c8d: Pull complete 
Digest: sha256:29e690dbe8e5c6920fc4bac2616ddb67bf593749ff218d755929bf14661cced0
Status: Downloaded newer image for farmos/farmos:7.x-1.x
 ---> 1397a8b028e6
Step 2/15 : ARG FARMOS_REPO=https://git.drupal.org/project/farm.git
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in 6c732a5a5833
Removing intermediate container 6c732a5a5833
 ---> 4abd43877bc2
Step 3/15 : ARG FARMOS_BRANCH=7.x-1.x
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in 1406a8dd1471
Removing intermediate container 1406a8dd1471
 ---> 8af9a5516fbf
Step 4/15 : ARG WWW_DATA_ID=1000
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in bca14ea09539
Removing intermediate container bca14ea09539
 ---> 63458629f5d6
Step 5/15 : RUN usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in 7a266d757fa9
standard_init_linux.go:219: exec user process caused: exec format error
The command '/bin/sh -c usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data' returned a non-zero code: 1
```

The normal docker image builds on top of a drupal:7 build for arm64.

1. Is there any reason for the dev image to be built differently, from the 7.x-1.x?
Prob would be good to build all images from source (keeping system config layers in cache), and tagging the :dev image building from the branch 7.x-1.x, and then

2. where can I find the pipeline definitions for the docker builds? I assume that you're not doing it from github?

3. would it be possible to build images for other (most used) arches to docker?

thanks

**Comment by mstenta** [comment]
> `The command '/bin/sh -c usermod -u ${WWW_DATA_ID} www-data && groupmod -g ${WWW_DATA_ID} www-data' returned a non-zero code: 1`

Ah interesting... this is actually a recent change that was made to the `7.x-1.x-dev` image: #384 

So for some reason those commands don't work with the ARM64 build process. Specifically, this is what's failing:

```
usermod -u 1000 www-data
groupmod -g 1000 www-data
```

You could comment that line out in your `docker/dev/Dockerfile` and try rebuilding. I bet that will work... I'm not sure why it's causing a problem in the first place though.

> where can I find the pipeline definitions for the docker builds? I assume that you're not doing it from github?

We have Docker Hub configured to watch the GitHub repository and automatically build the images whenever a commit is pushed.

> would it be possible to build images for other (most used) arches to docker?

That would be great! I've never gone down that road before with Docker Hub.

**Comment by raleonardo** [comment]
I commented the lines you mention, but I guess all the run commands are failing, being the next one in line:
```
raleonardo@rockpi-4b:~/farmOS$ docker build -t farmos/farmos/dev farmOS/docker/dev/
Sending build context to Docker daemon   5.12kB
Step 1/13 : FROM farmos/farmos:7.x-1.x
 ---> 1397a8b028e6
Step 2/13 : ARG FARMOS_REPO=https://git.drupal.org/project/farm.git
 ---> Using cache
 ---> 4abd43877bc2
Step 3/13 : ARG FARMOS_BRANCH=7.x-1.x
 ---> Using cache
 ---> 8af9a5516fbf
Step 4/13 : RUN yes | pecl install xdebug-2.9.8 	  && echo "zend_extension=$(find /usr/local/lib/php/extensions/ -name xdebug.so)" > /usr/local/etc/php/conf.d/xdebug.ini     && echo "xdebug.remote_enable=on" >> /usr/local/etc/php/conf.d/xdebug.ini     && echo "xdebug.remote_autostart=off" >> /usr/local/etc/php/conf.d/xdebug.ini
 ---> [Warning] The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
 ---> Running in e80d47d58f77
standard_init_linux.go:219: exec user process caused: exec format error
The command '/bin/sh -c yes | pecl install xdebug-2.9.8 	  && echo "zend_extension=$(find /usr/local/lib/php/extensions/ -name xdebug.so)" > /usr/local/etc/php/conf.d/xdebug.ini     && echo "xdebug.remote_enable=on" >> /usr/local/etc/php/conf.d/xdebug.ini     && echo "xdebug.remote_autostart=off" >> /usr/local/etc/php/conf.d/xdebug.ini' returned a non-zero code: 1
```
IMHO the run commands are being instantiated on amd64 layers (from farmos/farmos:7.x-1.x) from an arm64 machine within the docker build.


> That would be great! I've never gone down that road before with Docker Hub.
Cool! I found this documentation:
https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/

let me know if I can help on this.

**Comment by mstenta** [comment]
Hmm yea looks like it will require more digging...

> let me know if I can help on this.

This isn't currently on my priority list - but feel free to outline a plan and I'd be happy to provide feedback!

**Comment by raleonardo** [comment]
@mstenta 
I forked farmos and wrote a [draft](https://github.com/raleonardo/farmOS/blob/7.x-1.7.12/.github/workflows/main.yml) for a workflow that builds (amd64 + arm64) every time a new tag is pushed from the 7.x-1.x branch, and pushes the docker images with the corresponding tag to [dockerhub](https://hub.docker.com/repository/docker/raleonardo/farmos/tags?page=1&ordering=last_updated).

I know that this would entail you changing the complete CI from Docker to github, yet you might want to take a look :)

TODO: I couldn't get the docker layers caching between builds yet. it would speed up consecutive builds

**Comment by mstenta** [comment]
Thank you @raleonardo ! I would love to move to using GitHub Actions to build our images (and then push to Docker Hub). The Docker Hub build options leave a lot to be desired. AND this would allow us to build for ARM!

This isn't a super high priority for us right now (pushing on farmOS 2.x) but I'd like to come back and revisit this as soon as possible...

I'm thinking it would make sense to integrate into our existing workflows in `2.x` potentially... and maybe only after automated tests pass too. TBD...

**Comment by mstenta** [comment]
> I'm thinking it would make sense to integrate into our existing workflows in `2.x` potentially... and maybe only after automated tests pass too. TBD...

FYI discussing this a bit with @symbioquine in chat right now - for future reference. :-)

Starting here: https://irc.farmos.org/bot/log/farmOS/2021-03-11#T61789

**Comment by mstenta** [comment]
I started a new issue for the GitHub Action refactoring that would be required: https://www.drupal.org/project/farm/issues/3203129

Assuming we can figure that out, it seems like this would be an easy addition to it! Perhaps we should close this issue and make `buildx` and multi-architecture builds a core part of that issue?

**Comment by raleonardo** [comment]
Cool! Yeah sure! (close this one)
Are the drupal issues mirroring the github ones? 

Off topic: I'd like to make a feature request for new images (config including setup for staying behind proxy and config setup for accepting config via ENV (making it more K8 friendly)).
Should I post this in drupal/github issues?

thanks!

**Comment by raleonardo** [comment]
After reading the irc chat shortly, and yet so shortly (some ideas):

when I was using CI/CD in gitlab, we build and push the intermediate images into gitlabs private registry,so we can cache layers between builds and cache the images between tests in the pipeline, tagging them and pushing again as they pass them. as final stage you could push them into docker hub.

Yeah I know the 'local' registry might cost a bit extra, depending on usage, but should be fine just to keep the last stable, and the last dev for each arch. oh times 2 (1.x and 2.x)

**Comment by mstenta** [comment]
> Cool! Yeah sure! (close this one)

Great! Closing... :-)

> Are the drupal issues mirroring the github ones?

Not mirrored - but drupal.org has always been the "canonical" farmOS core issue queue. The GitHub repo mirror and issue queue came second, and has caused confusion because of this. I thought about shutting it down entirely, but we decided to instead try to use it only for bug reports, and guide people to the forum instead (and if they aren't sure which repo an issue belongs in): #374 

But there are still a bunch of open threads on here (now there is one fewer). One step at a time... :-)

> Off topic: I'd like to make a feature request for new images (config including setup for staying behind proxy and config setup for accepting config via ENV (making it more K8 friendly)).
> Should I post this in drupal/github issues?

Perhaps it makes sense to start a forum topic with the proposal? Then, once there is a concrete plan, we can create an official issue to track it.

> Yeah I know the 'local' registry might cost a bit extra, depending on usage, but should be fine just to keep the last stable, and the last dev for each arch. oh times 2 (1.x and 2.x)

I think it's free for public repos (?) - but agreed we can minimize it if we need to.

Re: "1.x and 2.x" I think I'd prefer to start with 2.x, but leave 1.x as-is... and then consider moving that as well in a follow-up. But honestly I'm eager to deprecate 1.x as soon as possible. :-)

**Comment by mstenta** [comment]
Quick update: we are now using GitHub Actions to build the images that we push to Docker Hub! #580 :tada: 

This should clear the way to adding an ARM64 image build process. I opened a new dedicated issue for that here: https://www.drupal.org/project/farm/issues/3322762
